src/graft_dag.cpp

```diff
@@ -0,0 +1,487 @@
+  // insert node into subsplit map.
+  SafeInsert(subsplit_to_id_, node_subsplit, node_id);
+}
+
+void GraftDAG::CreateGraftEdge(const size_t parent_id, const size_t child_id) {
+  // find relationship between nodes.
+  Bitset parent_subsplit = GetDAGNode(parent_id)->GetBitset();
+  Bitset child_subsplit = GetDAGNode(child_id)->GetBitset();
+  bool is_left =
+      !Bitset::SubsplitIsChildOfWhichParentClade(parent_subsplit, child_subsplit);
+  CreateGraftEdge(parent_id, child_id, is_left);
+}
+
+void GraftDAG::CreateGraftEdge(const size_t parent_id, const size_t child_id,
+                               const bool is_left) {
+  // Add edge to edge list.
+  Assert(ContainsNode(parent_id), "Node with the given parent_id does not exist.");
+  Assert(ContainsNode(child_id), "Node with the given child_id does not exist.");
+  auto parent_node = GetDAGNode(parent_id);
+  auto child_node = GetDAGNode(child_id);
+
+  Bitset parent_bitset = parent_node->GetBitset();
+  Bitset child_bitset = child_node->GetBitset();
+  Bitset edge_bitset = Bitset::PCSP(parent_bitset, child_bitset);
+  // Add edge to grafted edges.
+  bool edge_does_not_exist = !ContainsEdge(parent_id, child_id);
+  if (edge_does_not_exist) {
+    // std::cout << "** # Adding Edge to Graft (" << parent_id << ", " << child_id <<
+    // "): " << edge_bitset.EdgeToString() << std::endl;
+    Bitset edge_bitset = Bitset::PCSP(parent_bitset, child_bitset);
+    graft_edges_.insert(
+        std::make_pair<SizePair, size_t>({parent_id, child_id}, EdgeCount()));
+    // Add edge to individual SubsplitDAGNodes (only adds them to graft nodes).
+    bool parent_node_is_in_dag = !IsNodeFromHost(parent_id);
+    bool child_node_is_in_dag = !IsNodeFromHost(child_id);
+    if (is_left) {
+      if (!parent_node_is_in_dag) {
+        parent_node->AddLeftLeafward(child_node->Id());
+      }
+      if (!child_node_is_in_dag) {
+        child_node->AddLeftRootward(parent_node->Id());
+      }
+    } else {
+      if (!parent_node_is_in_dag) {
+        parent_node->AddRightLeafward(child_node->Id());
+      }
+      if (!child_node_is_in_dag) {
+        child_node->AddRightRootward(parent_node->Id());
+      }
+    }
+  }
+}
+
+SubsplitDAG::ModificationResult GraftDAG::AddGraftNodePair(const NNIOperation &nni) {
+  return AddGraftNodePair(nni);
+}
+
+SubsplitDAG::ModificationResult GraftDAG::AddGraftNodePair(
+    const Bitset &parent_subsplit, const Bitset &child_subsplit) {
+  // Initialize results.
+  SubsplitDAG::ModificationResult results;
+  // Assert that at least one of the nodes are not in the host DAG or graft DAG.
+  const bool parent_is_new =
+      !(host_dag_.ContainsNode(parent_subsplit) || ContainsGraftNode(parent_subsplit));
+  const bool child_is_new =
+      !(host_dag_.ContainsNode(child_subsplit) || ContainsGraftNode(child_subsplit));
+  // Assert that parent and child are a valid node pair to add to DAG.
+  if (host_dag_.IsValidAddNodePair(parent_subsplit, child_subsplit) == false) {
```

Comment:
I suggest putting these sort of checks first.

Reply:
Moved.